File: Java/Node.java
package net.gamepickle;

/**
 * Created by Joseph on 29/04/2017.
 */
public class Node {
    public double x;
    public double y;

    public Node(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double dot(Node node){
        // Dot Product between this vector and the passed one
        return x*node.x + y*node.y;
    }

    public Node normal() {
        // The normal of this vector
        return new Node(-1*y, x);
    }

    public Node normalize() {
        // The unit vector of this one
        double d = Math.sqrt(x*x+y*y);
        if(d == 0) {
            d = 1;
        }
        return new Node(x/d,y/d);
    }

}

File: Python/Node.py
import math

class Node(object):
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def dot(self, node):
        return self.x*node.x + self.y*node.y

    def normal(self):
        return Node(-self.y, self.x)

    def normalize(self):
        d = math.sqrt(self.x**2 + self.y**2)
        if d == 0:
            d = 1
        return Node(self.x/d, self.y/d)
